Operator.compose should allow indices

molar-volume opened the pull request:
resolves #1088 
resolves #1143 

This PR is created to verify the solution for expanding operator in #1088 (permute part of issue is not addressed yet)

The method ```expand``` for expanding the operator with ```identity``` is defined in OperatorBase,
because it enhances ```compose``` (defined for OperatorBase),
and uses ```tensor``` (defined for OperatorBase), and new method ```identity``` (abstract in OperatorBase).

```identity``` is implemented for each subclass of 'OperatorBase',

but I am unsure of implementation for DictStateFn and VectorStateFn (instead of identity, ```zeros``` are used).

The new features are demonstrated in```test_expand_on_state_fn``` and ```test_compose_op_of_different_dim```

### Summary
Progress for #1088 and solution for #1143 

### Details and comments

Cryoris commented:
Thanks for the commitment @molar-volume! I think we need to clarify what #1088 should do and how we could do it. The idea was that you could pass concrete indices (not only ops of different sizes) and e.g. do
```python
op = (X ^ 5)
op.compose( Z ^ Y, qubits=[0, 3], inplace=True)
print(op)  # ComposedOp( [X ^ X ^ X ^ X ^ X], [Z ^ I ^ I ^ Y ^ I] )
```
So it is closer to the same method of the quantum circuit, see [`QuantumCircuit.compose`](https://github.com/Qiskit/qiskit-terra/blob/2bd4940882fd46e8d1463d7ebb65c831ddd55b8a/qiskit/circuit/quantumcircuit.py#L546). (You could also add the `front` feature if you want! 🙂 ).

The approach in here does only allow for consecutive qubit indices (e.g. `[1,2,3]`) not for others (such as `[0, 3]`) if I understand it correctly. I have two concerns about the methods you added:
* `expand` already exists in the `quantum_info` operators but has a different signature. There you can call `Operator.expand(other_operator_of_same_dim)`, I think we should keep the same syntax if possible and therefore not call it `expand` here. If this turns out to be necessary we could possibly call it `expand_to_dim` or `pad_with_identities` or so.
* Adding an `identity` method is maybe not the best approach since it is not clear to me what you would expect upon calling `CircuitOp.identity`? On a circuit this adds an identity gate, here it does something different. You could hide this inside `pad_with_identities`.

A possible approach would be to add a `pad_with_identities` (or whatever you want to call it) method looking like
```python
def pad_with_identities(self, new_dim, qargs=None):
    """Extend this operator to ``new_dim`` where this operator acts on ``qargs``.

    Args:
        new_dim: The new dimension, must be larger equal than ``self.num_qubits``.
        qargs: The qubit indices this operator acts on, must be a list of length ``self.num_qubits``.
    """
```
And then in `compose` you can use the above to expand `other` to the correct size and compose it. What do you think?

molar-volume commented:
Hi @Cryoris thank you for the feedback !
Yes, let’s rename ```expand``` to ```expand_to_dim```.

Now ```identity``` (or we should rather name it ```identity_operator```) is OperatorBase with ```primitive``` corresponding to identity matrix, so for CircuitOp, it equals to ```(I ^ num_qubits).to_circuit_op()``` .

I think it has some advantages, for example you can define ```expand_to_dim``` for all subclasses in OperatorBase as
```python
def expand_to_dim(self, num_qubits: int) -> OperatorBase:
	return self.tensor(self.identity_operator(num_qubits))
```

I understand that we need a more general method with indices, just like you defined
```python 
def pad_with_identities(self, new_dim, qargs=None)
``` 

For the ```CircuitOp```, that functionality can be already provided using ```permute``` (as suggested by @dongreenberg )
```python
circ_op = (X ^ Y).to_circuit_op().permute([0,2]) @ (X ^ Y ^ Z ^ X).to_circuit_op()

print(circ_op.to_circuit())
     ┌───┐     
q_0: ┤ X ├─────
     ├───┤┌───┐
q_1: ┤ Z ├┤ Y ├
     ├───┤└───┘
q_2: ┤ Y ├─────
     ├───┤┌───┐
q_3: ┤ X ├┤ X ├
     └───┘└───┘
```

I think I know how to implement ```permute``` for ```PauliOp```, but do you have any suggestion how to do it for ```MatrixOp```?

molar-volume commented:
Permute implemented for PauliOp and MatrixOp as well. 
```python
import numpy as np
from qiskit.aqua.operators import MatrixOp, X, Y, Z, I, CircuitOp

pauli_op = (X ^ Z).permute([0, 2]) @ (X ^ Y ^ Z ^ X)

matrix_op = (X ^ Z).to_matrix_op().permute([0, 2]) @ (X ^ Y ^ Z ^ X).to_matrix_op()

circuit_op = (X ^ Z).to_circuit_op().permute([0, 2]) @ (X ^ Y ^ Z ^ X).to_circuit_op()
print(circuit_op.to_circuit())

circ_eq_pauli = np.allclose(circuit_op.to_matrix(), pauli_op.to_matrix())  # circ_eq_pauli: True
circ_eq_matrix = np.allclose(circuit_op.to_matrix(), matrix_op.to_matrix())  # circ_eq_matrix: True

     ┌───┐┌───┐
q_0: ┤ X ├┤ I ├
     ├───┤├───┤
q_1: ┤ Z ├┤ Z ├
     ├───┤└───┘
q_2: ┤ Y ├─────
     ├───┤┌───┐
q_3: ┤ X ├┤ X ├
     └───┘└───┘
```
Do you want to provide compose with indices (or ```permute```) for ListOps as well @Cryoris ?

molar-volume commented:
The implementation of permute for ListOps might go in line with MatrixOp.permute

woodsp-ibm commented:
Is there are good reason for the rename of `identity` to `identity_operator`? Since this is a breaking change, if we needed to do this, then we ought to be keep `identity` still (which would call over to the new function) but issue a deprecation warning with an appropriate message to use the method by its new name.

molar-volume commented:
@woodsp-ibm ```identity``` in OperatorBase was defined by me as part of this PR, I changed it to ```identity_operator```, because @Cryoris pointed out that it is unclear what to expect from calling ```CircuitOp.identity```

molar-volume commented:
after reading #1086 I note that expand_to_dim will be reimplemented for SummedOp and ComposedOp, because e.g ```SummedOp.num_qubits``` is supposed to return the ```List[int]```

woodsp-ibm commented:
Ah ok, identity is a new method that was added by this PR. so you just renamed your new method. From my perspective I am not sure that the name change does so much for me since they are all operators according to OperatorBase. Maybe the method anyway should be some class or even module level function to return a specific operator of said type i.e. an identity in this case. I do not know why it would be an instance method since it has no real need of self (or am I mistaken again) - unless it were an identity spanning the same number of qubits as the operator in self in which case I could somehow see it.

molar-volume commented:
@woodsp-ibm ```self.identity_operator``` is called in method ```expand_to_dim```, which is defined in OperatorBase.
So based on the ```self``` the specific implementation of ```identity_operator```.

woodsp-ibm commented:
Yes, I see some implementations of identity_operator do use self (when they need to copy across a property like is_measurement) while some do not. Do you see this as function that would be called in general by a user of the API - I guess you can if you want an instance that is an identity across that number of qubits - but of course with it being tied to self one needs to do that off an existing object which just seems weird. Otherwise its just a really some helper method to facilitate the one expand_to_dim implementation in the base class rather than have it implemented in each of the classes.

Cryoris commented:
I agree with @woodsp-ibm, `identity_operator` is more a helper for `expand_to_dim` than a method, it doesn't implement enough logic IMO. Could you just put the code inside `expand_to_dim`, since its only a few lines anyways?

molar-volume commented:
Right, originally I defined it as ```python
@staticmethod
    def identity_operator(num_qubits: int)
```
but then I needed self anyway for StateFns, so I redefined it as class method. As static method, you could use it as e.g. PauliOp.identity_operator

Cryoris commented:
Yeah, but I don't think its a good idea to add an `identity_operator` method. This seems like a very specific functionality and it doesn't add so much logic. I think it is better to don't overpopulate the Operators with too many methods 🙂 If later we find, that this is actually a much needed functionality we can still add it.

molar-volume commented:
@Cryoris OK I will, I can do some optimisations then, instead of using ```tensor``` all the time.

molar-volume commented:
@Cryoris @woodsp-ibm 
I think this PR is ready, so if you have some comments or hints for improvement, I will do my best 👍 
What was achieved:

1) ```expand_with_identity``` implemented for each subclass of OperatorBase
2) ```permute``` implemented for each subclass of OperatorBase
3) ```compose``` signature changed to accept indices for permutation of self and other
4) composing of operators with different dimension enabled (automatic expansion of the smaller operator)
5) minor refactoring to simplify the new features
6) unit tests for the new features

woodsp-ibm commented:
@molar-volume Thanks - will take a look. Sorry we have been rather busy with the release this week.

Cryoris commented on qiskit/aqua/operators/state_fns/dict_state_fn.py:
```diff
@@ -109,6 +110,39 @@ def adjoint(self) -> OperatorBase:
+        return DictStateFn(new_dict, coeff=self.coeff, is_measurement=self.is_measurement)
+
+    def to_vector_state_fn(self) -> 'VectorStateFn':  # type: ignore
+        r"""
+        Creates the equivalent state function of type VectorStateFn.
```
```suggestion
        """Creates the equivalent state function of type VectorStateFn.
```

Cryoris commented on qiskit/aqua/operators/state_fns/circuit_state_fn.py:
```diff
@@ -365,5 +373,5 @@ def permute(self, permutation: List[int]) -> 'CircuitStateFn':
             A new CircuitStateFn containing the permuted circuit.
         """
-        new_qc = QuantumCircuit(self.num_qubits).compose(self.primitive, qubits=permutation)
+        new_qc = QuantumCircuit(max(permutation) + 1).compose(self.primitive, qubits=permutation)
```
This suggests that I can do 
```python
fn = One.to_circuit_op()  # CircuitStateFn
fn.permute([10000])
```
Is this the case? Is that also true for all operators?

molar-volume replied:
Yes you can do it for all operators.
Sorry I missed this comment earlier.

Cryoris commented on qiskit/aqua/operators/state_fns/circuit_state_fn.py:
```diff
@@ -354,6 +359,9 @@ def reduce(self) -> OperatorBase:
                     del self.primitive.data[i]
         return self
 
+    def expand_with_identity(self, num_qubits: int) -> 'CircuitStateFn':
+        return self.permute(list(range(num_qubits, num_qubits + self.num_qubits)))
```
Nice one, but a bit confusing to read. Could you add a comment? 
Also: Is this the order? Meaning should the identities be padded on top or bottom of the circuit?

molar-volume replied:
Sure, good point, it deserves commenting, that we are reversing qiskit endinanness here, just like in ```tensor``` method in CircuitOp.
So this is just ```self.tensor(identity)```.

Cryoris commented on qiskit/aqua/operators/primitive_ops/pauli_op.py:
```diff
@@ -98,9 +101,35 @@ def tensor(self, other: OperatorBase) -> OperatorBase:
-    def compose(self, other: OperatorBase) -> OperatorBase:
-        other = self._check_zero_for_composition_and_expand(other)
+    def permute(self, permutation: List[int] = None) -> 'PauliOp':
+        """ Permutes the sequence of Pauli matrices.
```
```suggestion
        """Permutes the sequence of Pauli matrices.
```

Cryoris commented on qiskit/aqua/operators/primitive_ops/matrix_op.py:
```diff
@@ -108,21 +110,62 @@ def equals(self, other: OperatorBase) -> bool:
+        return super(MatrixOp, self).compose(other)
+
+    def permute(self, permutation: List[int] = None) -> 'MatrixOp':
+        """ Creates a new MatrixOp that acts on the permuted qubits.
```
```suggestion
        """Creates a new MatrixOp that acts on the permuted qubits.
```

Cryoris commented on qiskit/aqua/operators/primitive_ops/matrix_op.py:
```diff
@@ -108,21 +110,62 @@ def equals(self, other: OperatorBase) -> bool:
+                permute_self: List[int] = None,
+                permute_other: List[int] = None) -> OperatorBase:
 
+        self, other = self._check_zero_for_composition_and_expand(other,  # type: ignore
```
Could you add a comment that this does the proper resorting? It's not very obvious from the method name

molar-volume replied:
maybe we can rename the method to ```_expand_the_shorter_and_permute```?
check_zero functionality only expands the shorter operator if it's Zero

Cryoris commented on qiskit/aqua/operators/evolutions/evolved_op.py:
```diff
@@ -88,9 +89,19 @@ def tensor(self, other: OperatorBase) -> OperatorBase:
 
-    def compose(self, other: OperatorBase) -> OperatorBase:
-        other = self._check_zero_for_composition_and_expand(other)
+    def expand_with_identity(self, num_qubits: int) -> 'OperatorBase':
```
Could we make these private until we find a user-facing use-case? 
```suggestion
    def _expand_with_identity(self, num_qubits: int) -> 'OperatorBase':
```

molar-volume replied:
Sure 👍

molar-volume commented on qiskit/aqua/operators/list_ops/composed_op.py:
```diff
@@ -65,7 +65,12 @@ def distributive(self) -> bool:
 
-    def compose(self, other: OperatorBase) -> OperatorBase:
+    def compose(self, other: OperatorBase,
+                permute_self: List[int] = None,
+                permute_other: List[int] = None) -> OperatorBase:
```
permuting the first Operator is enabled in ```Operator``` using the ```front=True```, like this:
```python
Operator('XYZ').compose(Operator('XY'), qargs=[0,2], front=True) === Operator('XIY').compose(Operator('XYZ'))
```
So we can do it that way if you want.
IMO approach with two set of indices is more intuitive.

Cryoris replied:
Yeah it might be more intuitive, but consistency is a bit more important I think as we are trying to get on a path to unify the operators. Thanks! 👍

molar-volume commented:
Changes based on @Cryoris 's code review:
1) To avoid using ```sympy```, custom method implemented in arithmetic module.
2) Signature of compose changed - accepts a single list of indices for permutation of other operator.
3) Front feature for compose - if parameter ```front==True```, return ```other.compose(self)```, but only after expansion and permutation is performed (consistent with Operator in quantum_info)
4) Reimplemented ```VectorStateFn.permute``` for better performance
5) ```expand_with_identity``` made private and renamed to ```_expand_dim``` (former name was misleading, because for DictStateFn and VectorStateFn, ```zeros``` are used for expansion)
6) ```_check_for_zero_composition_and_expand``` renamed to ```_expand_shorter_operator_and_permute```, which is more self-explanatory (```_check_for_zero_composition``` feature only expands the Zero operator)
7) added comment to explain ```CircuitStateFn._expand_dim```
8) modified unit tests to test new features

woodsp-ibm commented on qiskit/aqua/operators/evolutions/evolved_op.py:
```diff
@@ -88,9 +89,20 @@ def tensor(self, other: OperatorBase) -> OperatorBase:
 
-    def compose(self, other: OperatorBase) -> OperatorBase:
-        other = self._check_zero_for_composition_and_expand(other)
+    def _expand_dim(self, num_qubits: int) -> 'OperatorBase':
```
I think the return type here can be `OperatorBase` directly right and does not need to be a string (a string is normally only needed for some forward ref. or ref to itself. There is another location just below and the same comment in a couple of other modules too where I do not think string is needed.

woodsp-ibm commented on qiskit/aqua/utils/arithmetic.py:
```diff
@@ -95,3 +95,33 @@ def next_power_of_2_base(n):
     return base
+
+
+def transpositions(permutation):
```
Could we have the typehint on the parameter here and Args in docstring describing it. Thx.

Cryoris commented on qiskit/aqua/operators/state_fns/state_fn.py:
```diff
@@ -190,24 +206,20 @@ def tensorpower(self, other: int) -> Union[OperatorBase, int]:
 
-    def _check_zero_for_composition_and_expand(self, other: OperatorBase) \
+    def _expand_shorter_operator_and_permute(self, other: OperatorBase,
+                                             permutation: List[int] = None) \
```
```suggestion
                                             permutation: Optional[List[int]] = None) \
```

Cryoris commented on qiskit/aqua/operators/state_fns/vector_state_fn.py:
```diff
@@ -78,6 +81,51 @@ def adjoint(self) -> OperatorBase:
+                             is_measurement=self.is_measurement)
+
+    def to_dict_fn(self) -> StateFn:
+        r"""
+        Creates the equivalent state function of type DictStateFn.
```
```suggestion
        """Creates the equivalent state function of type DictStateFn.
```

Cryoris commented on qiskit/aqua/utils/arithmetic.py:
```diff
@@ -95,3 +96,40 @@ def next_power_of_2_base(n):
+            first = x[0]
+            for y in x[len_x - 1:0:-1]:
+                res.append((first, y))
+    return res  # type: ignore
```
This seems like it shouldn't be ignored -- maybe the type hint it not quite correct if mypy complains

molar-volume replied:
mypy expects ```List[Tuple[int,any]]```, but in the algorithm, we always append ```Tuple[int,int]```, so I thought it is better to ignore mypy here, then provide the confused type hint

woodsp-ibm replied:
I think if you change the line `res.append(tuple(x))` in this below to `res.append((x[0], x[1]))` then mypy will see the types as consistent and you can remove the ignore
```
    for x in cyclic_form:
        len_x = len(x)
        if len_x == 2:
            res.append(tuple(x))
```

molar-volume replied:
@woodsp-ibm yes it helped, thanks

Cryoris reviewed (changes requested):
Hi @molar-volume,

Thanks a lot for the updates, this is going in the right direction. I still have some comments, mainly about the fact that all elements in the opflow should be immutable. Hence you shouldn't overwrite `self` anywhere. Also some type hints are left to adjust, but otherwise looks good!

molar-volume commented on qiskit/aqua/operators/evolutions/evolved_op.py:
```diff
@@ -88,9 +88,20 @@ def tensor(self, other: OperatorBase) -> OperatorBase:
+    def compose(self, other: OperatorBase,
+                permutation: List[int] = None, front: bool = False) -> OperatorBase:
+
+        self, other = self._expand_shorter_operator_and_permute(other,  # type: ignore
```
Inside the method, ```self``` is treated as local variable. 
Replacing ```self``` will not replace the original instance.

Cryoris replied:
Oh can you rename the variable then? Calling a variable `self` seems very confusing (example: me 😛 )

molar-volume commented on qiskit/aqua/operators/list_ops/list_op.py:
```diff
@@ -201,7 +204,52 @@ def tensorpower(self, other: int) -> Union[OperatorBase, int]:
+        Raises:
+            AquaError: if indices do not define a new index for each qubit.
+        """
+        new_self = self
```
I think this is not necessary, see the comment above.

molar-volume commented on qiskit/aqua/operators/evolutions/evolved_op.py:
```diff
@@ -88,9 +88,20 @@ def tensor(self, other: OperatorBase) -> OperatorBase:
+        from qiskit.aqua.operators import I
+        return self.tensor(I ^ num_qubits)
 
+    def permute(self, permutation: List[int]) -> OperatorBase:
```
Hi, the docstrings for ```compose``` and ```permute``` are provided in OperatorBase,
and e.g. MatrixOp inherits the docstring documentation.https://qiskit.org/documentation/stubs/qiskit.aqua.operators.primitive_ops.MatrixOp.compose.html#qiskit.aqua.operators.primitive_ops.MatrixOp.compose

I am not sure if it is necessary to override docstrings for all ```compose``` methods (and for ```permute``` as well), because we were not doing it so far.

But if you are sure that it is the right move, I can do it.

molar-volume commented on qiskit/aqua/operators/list_ops/list_op.py:
```diff
@@ -201,7 +204,52 @@ def tensorpower(self, other: int) -> Union[OperatorBase, int]:
+        new_self = self
+        circuit_size = max(permutation) + 1
+
+        if self.num_qubits != len(permutation):
```
Hi, for ListOp, the ```num_qubits``` is defined in the same way as in ```ComposedOp``` and ```SummedOp```, and you can also initialize ```ComposedOp([X, X^X])``` or ```SummedOp([X, X^X])```, so ListOp does not seem to be exceptional.

molar-volume replied:
@Cryoris I just checked the incoming changes from https://github.com/Qiskit/qiskit-aqua/pull/1189 and if the operators hold different number of qubits, the ```ListOp.num_qubits``` will raise the exception.

To treat the ```ComposedOp([X, X^X])``` and ```SummedOp(X, X^X)]```, maybe we can  add ```expand_dim``` to ```__init__``` methods of ```ComposedOp``` and ```SummedOp```?

molar-volume replied:
@Cryoris OK, I will catch the exception, and raise the new exception with different message

molar-volume commented on qiskit/aqua/operators/evolutions/evolved_op.py:
```diff
@@ -88,9 +88,20 @@ def tensor(self, other: OperatorBase) -> OperatorBase:
+    def compose(self, other: OperatorBase,
+                permutation: List[int] = None, front: bool = False) -> OperatorBase:
+
+        self, other = self._expand_shorter_operator_and_permute(other,  # type: ignore
```
```self``` is ```EvolvedOp```, so mypy complained (but I will change ```self``` to ```new_self``` and delete # type: ignore)

molar-volume commented on qiskit/aqua/operators/operator_base.py:
```diff
@@ -491,6 +517,25 @@ def _get_param_dict_for_index(unrolled_dict: Dict[ParameterExpression, List[Numb
+            elif other.num_qubits < self.num_qubits:
+                other = other._expand_dim(self.num_qubits - other.num_qubits)
+            elif other.num_qubits > self.num_qubits:
+                # type: ignore
```
right, it is unnecessary

molar-volume commented:
Hi @Cryoris 🙂  thank you for your review and for all the comments.

molar-volume commented on qiskit/aqua/operators/primitive_ops/primitive_op.py:
```diff
@@ -155,21 +155,18 @@ def tensorpower(self, other: int) -> Union[OperatorBase, int]:
+                new_oplist = [comp_with_first] + other.oplist[1:]
+                return ComposedOp(new_oplist, coeff=other.coeff)
+            return ComposedOp([self] + other.oplist, coeff=other.coeff)  # type: ignore
+
```
Sorry I missed this comment before.
Yes I mixed it, but I think that PR is already merged here, so this should be resolved.

molar-volume commented on qiskit/aqua/operators/state_fns/dict_state_fn.py:
```diff
@@ -109,6 +110,39 @@ def adjoint(self) -> OperatorBase:
+        new_dict = {key + pad: value for key, value in self.primitive.items()}
+        return DictStateFn(new_dict, coeff=self.coeff, is_measurement=self.is_measurement)
+
+    def to_vector_state_fn(self) -> 'VectorStateFn':  # type: ignore
```
OK, I removed this ```to_vector_state_fn``` because ```to_matrix_op``` is already imlemented in ```state_fn```

Cryoris commented on qiskit/aqua/operators/operator_base.py:
```diff
@@ -498,6 +524,24 @@ def _get_param_dict_for_index(unrolled_dict: Dict[ParameterExpression, List[Numb
         return {k: v[i] for (k, v) in unrolled_dict.items()}
 
+    def _expand_shorter_operator_and_permute(self, other: 'OperatorBase',
+                                             permutation: List[int] = None) \
```
```suggestion
                                             permutation: Optional[List[int]] = None) \
```

Cryoris replied:
This isn't resolved, could you please update this type hint? If `None` is allowed as value the type hint must include that. Either via `Union[List[int], None] = None` or `Optional[List[int]] = None`

molar-volume replied:
@Cryoris it is resolved, the suggestion is outdated and in source code, there is already Optional[List[int]] = None

Cryoris commented on qiskit/aqua/operators/state_fns/vector_state_fn.py:
```diff
@@ -13,8 +13,9 @@
 
 
-from typing import Union, Set, Optional, Dict
+from typing import Union, Set, Optional, Dict, Listt
```
```suggestion
from typing import Union, Set, Optional, Dict, List
```

Cryoris reviewed (approved):
LGTM thanks for all the work!

woodsp-ibm commented:
I should add thanks for all your hard work and persistence to see this through. It has been a while, with a lot of discussion, so thanks again for your patience and perseverance in making this happen!

molar-volume commented:
My pleasure @woodsp-ibm ! I know the discussion was necessary and helped me a lot, so thanks for all your comments. Thanks @Cryoris !
